libtransmission/tr-buffer.h

```diff
@@ -264,7 +264,22 @@ class StackBuffer final
 
     virtual std::pair<value_type*, size_t> reserve_space(size_t n_bytes) override
     {
-        buf_.resize(end_pos_ + n_bytes);
+        if (auto const free_at_end = buf_.size() - end_pos_; free_at_end < n_bytes)
+        {
+            if (auto const total_free = begin_pos_ + free_at_end; total_free >= n_bytes)
+            {
+                // move data so that all free space is at the end
+                auto const size = this->size();
+                std::copy_n(data(), size, std::data(buf_));
```

Comment:
`std::copy_n()` [might not preserve element order](https://en.cppreference.com/w/cpp/algorithm/copy_n) if the source and destination ranges overlap.

> Overlap of ranges is formally permitted, but leads to unpredictable ordering of the results.

`std::copy()` must be used here.

Reply:
!

TIL. Thank you for that!